streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -756,7 +746,13 @@ public boolean maybePunctuateStreamTime() {
         if (timestamp == RecordQueue.NOT_KNOWN) {
             return false;
         } else {
-            return streamTimePunctuationQueue.mayPunctuate(timestamp, PunctuationType.STREAM_TIME, this);
+            final boolean punctuated = streamTimePunctuationQueue.mayPunctuate(timestamp, PunctuationType.STREAM_TIME, this);
+
+            if (punctuated) {
+                commitNeeded = true;
```

Comment:
Not sure I follow your comment here.. let me elaborate a bit on my logic:

We have two commits in places: commitAll (periodic) and maybeCommit (for user requested):

The latter checks

```
if (task.commitRequested() && task.commitNeeded()) 
                    task.commit();
```

While the former only checks:

```
if (task.commitNeeded()) 
                    task.commit();
```

I.e. the logic for the latter is that "only if user have requested, and it is indeed needed to commit": for example, if we have actually committed from the commit interval, and then user requested it as well, the second will be omitted.

I intentionally separated "commitRequest" (this is only set by user) and "commitNeeded" (this is determined by the library) because this way looks cleaner to me.

Reply:
So setting `commitNeeded` is a conservative approach, because we don't know what the user did within punctuation call? Might be better to set `commitNeeded` if user calls `context.forward` or `state.put()` -- not sure how hard this would be -- would also be out-of-scope for this PR. If we think it might be worth it, we should create a JIRA for this optimization.

Reply:
We still need to commit even if no records are processed: consider a topology which only contains a single source node, then no data processed at all, but we still want to commit so that we would not re-process them right?

Reply:
Fair enough. Thanks for pointing it out.